### PR TITLE
Blob info caching + mount/reuse

### PR DESCRIPTION
**Pull request by mtrmac**
This adds types.BlobInfoCache recording the encountered compressed/uncompressed digest relationships and `docker://` blob locations, and using them to:
- Mount blobs on `docker://` destinations instead of copying the data again
- Reuse local layers on `containers-storage:` destinations instead of pulling them
- Avoid blob decompression when converting from schema1 just to figure out the DiffID values

The blob info cache is stored in `/var/lib/containers/blob-info-cache-v1.boltdb` for EUID=0 users, and `$XDG_DATA_HOME/containers/blob-info-cache-v1.boltdb` for unprivileged users.

See individual commit messages for more details.

WIP items:
- <s>When `canSubstitute`, `dockerImageDestination` makes a mount request even within the same repo, which is unnecessary.</s>
- <s>Documentation of the `BlobInfoCache` interface and possibly other code</s>
- Tests for corner cases of the BoltDB code
- Logging on failures in the BoltDB code

### CODE REVIEW

**Comment by rhatdan** [comment]
@vrothberg @runcom PTAL

**Comment by rhatdan** [comment]
@nalind PTAL

**Comment by mtrmac** [comment]
Tests are failing here due to the API change, updated Skopeo is running the tests in https://github.com/containers/skopeo/pull/570 .

**Comment by mtrmac** [comment]
> The blob info cache is stored in `/var/lib/containers/blob-info-cache-v1.boltdb` for EUID=0 users, and `$XDG_DATA_HOME/containers/blob-info-cache-v1.boltdb` for unprivileged users.

This might not be the right thing to do - apparently rootless buildah runs in a user namespace, which causes it to see EUID==0, and try to write to `/var/lib/containers`.  That does not _break_ builds because any error causes the code to fall back to memory-only caches, but it makes the optimizations unavailable.

File: pkg/blobinfocache/boltdb.go [hunk 1]
```diff
@@ -0,0 +1,310 @@
...
+					}
+				}
+				if uncompressedDigestValue != primaryDigest {
+					res = bdc.appendReplacementCandidates(res, scopeBucket, uncompressedDigestValue)
+				}
+			}
+		}
+		return nil
+	}); err != nil { // Including os.IsNotExist(err)
+		return []types.BICReplacementCandidate{} // FIXME? Log?
```
Comment (rhatdan) [inline pkg/blobinfocache/boltdb.go]: Shouldn't we log everything except IsNotExist?

File: copy/copy.go [hunk 2]
```diff
@@ -502,28 +502,19 @@ type diffIDResult struct {
...
+
+	// If we already have the blob, and we don't need to compute the diffID, then we don't need to read it from the source.
+	if !diffIDIsNeeded {
+		reused, blobInfo, err := ic.c.dest.TryReusingBlob(ctx, srcInfo)
 		if err != nil {
-			return types.BlobInfo{}, "", errors.Wrapf(err, "Error reapplying blob %s at destination", srcInfo.Digest)
+			return types.BlobInfo{}, "", errors.Wrapf(err, "Error trying to reuse blob %s at destination", srcInfo.Digest)
+		}
+		if reused {
+			ic.c.Printf("Skipping fetch of repeat blob %s\n", srcInfo.Digest)
```
Comment (rhatdan) [inline copy/copy.go]: Can you reword this?

Skipping blob %s fetch, already exists?
Is this something we actually need to see, or would it be better to just see what is pulled?

**Comment by rhatdan** [comment]
@vrothberg @runcom @nalind @umohnani8 PTAL

**Comment by rhatdan** [comment]
@mtrmac Is this still WIP?  Can we get it vendored into skopeo so we can get tests to pass?

**Comment by mtrmac** [comment]
> @mtrmac Is this still WIP?

Yes, at least for documentation.

> Can we get it vendored into skopeo so we can get tests to pass?

Already exists in 
https://github.com/containers/skopeo/pull/570 .

File: pkg/blobinfocache/boltdb.go [hunk 3]
```diff
@@ -0,0 +1,310 @@
...
+		}
+		b, err = b.CreateBucketIfNotExists([]byte(uncompressed.String()))
+		if err != nil {
+			return err
+		}
+		if err := b.Put([]byte(anyDigest.String()), []byte{}); err != nil { // Possibly writing the same []byte{} presence marker again.
+			return err
+		}
+		return nil
+	}) // FIXME? Log error?
```
[the remark this answers is not in the document]
Comment (mtrmac) [reply pkg/blobinfocache/boltdb.go]: Yes, probably. That’s one of the WIP parts.

File: copy/copy.go (hunk 2 above)
Comment (mtrmac) [reply copy/copy.go]: (“fetch” in this message includes pushes.)

This can definitely be improved, at least `s/blob/layer/`, and maybe we don’t need the digests in the output at all. That’s all fairly independent from the functionality of this PR.

Comment (mtrmac) [reply copy/copy.go]: > (“fetch” in this message includes pushes.)

i.e. all of pull/push/copy go through this code path and may skip transferring a blob between transports here.

Comment (rhatdan) [reply copy/copy.go]: Does this just become noise on the screen, would the user care?  Seems to be more debugging information.

**Comment by rhatdan** [comment]
I'm willing to take it without Documentation, so we can move the testing along. Then open a separate PR for Docs

Why is the tests failing?

File: pkg/blobinfocache/boltdb.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (mtrmac) [reply pkg/blobinfocache/boltdb.go]: (The a bit tricky aspect to this is that failures using the cache are _never_ fatal (which might not be a good idea, to be fair), so a failure could result in dozens of identical or similar error messages per copy; we probably want to throttle them somehow, or maybe even mark the cache as broken after a few failures.)

**Comment by mtrmac** [comment]
Updated to avoid a no-op mount within a single repo, and added documentation to the interface/implementations; I can live with this being merged.

Tests fail here due to the API change, should eventually pass in https://github.com/containers/skopeo/pull/570 (and after merging this, a variant of https://github.com/containers/skopeo/pull/570 that vendors from `master` of this repo will need to be quickly merged).

**Comment by rhatdan** [comment]
LGTM
@giuseppe @vrothberg @nalind @runcom @umohnani8 PTAL

**Comment by runcom** [comment]
Reviewing this in a bit

File: copy/copy.go (hunk 2 above)
Comment (runcom) [reply copy/copy.go]: what about `Blob transfer skip` - more generic than fetch and more valid in the context of pushes

**Comment by runcom** [comment]
I had just a comment which we can live with for now/beginning of January.
This PR also passes our demo and code is pretty solid as well.

I'm gonna merge this, if you have anything against, please fill out issues/PRs

**Comment by runcom** [comment]
Tests in skopeo pass as well https://github.com/containers/skopeo/pull/570
